library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/Ac4Reader.java

```diff
@@ -0,0 +1,215 @@
+import com.google.android.exoplayer2.extractor.ExtractorOutput;
+import com.google.android.exoplayer2.extractor.TrackOutput;
+import com.google.android.exoplayer2.extractor.ts.TsPayloadReader.TrackIdGenerator;
+import com.google.android.exoplayer2.util.ParsableBitArray;
+import com.google.android.exoplayer2.util.ParsableByteArray;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+/**
+ * Parses a continuous AC-4 byte stream and extracts individual samples.
+ */
+public final class Ac4Reader implements ElementaryStreamReader {
+
+  @Documented
+  @Retention(RetentionPolicy.SOURCE)
+  @IntDef({STATE_FINDING_SYNC, STATE_READING_HEADER, STATE_READING_SAMPLE})
+  private @interface State {}
+
+  private static final int STATE_FINDING_SYNC = 0;
+  private static final int STATE_READING_HEADER = 1;
+  private static final int STATE_READING_SAMPLE = 2;
+
+  private final ParsableBitArray headerScratchBits;
+  private final ParsableByteArray headerScratchBytes;
+  private final String language;
+
+  private String trackFormatId;
+  private TrackOutput output;
+
+  @State private int state;
+  private int bytesRead;
+
+  // Used to find the header.
+  private boolean lastByteWasAC;
+  private boolean hasCRC;
+
+  // Used when parsing the header.
+  private long sampleDurationUs;
+  private Format format;
+  private int sampleSize;
+
+  // Used when reading the samples.
+  private long timeUs;
+
+  /**
+   * Constructs a new reader for AC-4 elementary streams.
+   */
+  public Ac4Reader() {
+    this(null);
+  }
+
+  /**
+   * Constructs a new reader for AC-4 elementary streams.
+   *
+   * @param language Track language.
+   */
+  public Ac4Reader(String language) {
+    headerScratchBits = new ParsableBitArray(new byte[Ac4Util.HEADER_SIZE_FOR_PARSER]);
+    headerScratchBytes = new ParsableByteArray(headerScratchBits.data);
+    state = STATE_FINDING_SYNC;
+    bytesRead = 0;
+    lastByteWasAC = false;
+    hasCRC = false;
+    this.language = language;
+  }
+
+  @Override
+  public void seek() {
+    state = STATE_FINDING_SYNC;
+    bytesRead = 0;
+    lastByteWasAC = false;
+    hasCRC = false;
+  }
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, TrackIdGenerator generator) {
+    generator.generateNewId();
+    trackFormatId = generator.getFormatId();
+    output = extractorOutput.track(generator.getTrackId(), C.TRACK_TYPE_AUDIO);
+  }
+
+  @Override
+  public void packetStarted(long pesTimeUs, @TsPayloadReader.Flags int flags) {
+    timeUs = pesTimeUs;
+  }
+
+  @Override
+  public void consume(ParsableByteArray data) {
+    while (data.bytesLeft() > 0) {
+      switch (state) {
+        case STATE_FINDING_SYNC:
+          if (skipToNextSync(data)) {
+            state = STATE_READING_HEADER;
+            headerScratchBytes.data[0] = (byte)0xAC;
+            headerScratchBytes.data[1] = 0x40;
```

Comment:
Nit: Maybe just `headerScratchBytes.data[1] = hasCRC ? 0x41 : 0x40;`?

Reply:
You are right. I'll update it.